Client/Application/AppDelegate.swift

```diff
@@ -319,92 +243,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
-            log.error("Custom URL schemes not available for validating")
-            return false
-        }
-
-        guard let scheme = components.scheme, urlSchemes.contains(scheme) else {
-            log.warning("Cannot handle \(components.scheme ?? "nil") URL scheme")
-            return false
-        }
-
-        guard let host = url.host else {
-            log.warning("Cannot handle nil URL host")
-            return false
-        }
-
-        let query = url.getQuery()
-
-        switch host {
-        case "open-url":
-            let url = query["url"]?.unescape() ?? ""
-            let isPrivate = NSString(string: query["private"] ?? "false").boolValue
-
-            let params = LaunchParams(url: URL(string: url), isPrivate: isPrivate)
-
-            if application.applicationState == .active {
-                // If we are active then we can ask the BVC to open the new tab right away.
-                // Otherwise, we remember the URL and we open it in applicationDidBecomeActive.
-                launchFromURL(params)
-            } else {
-                openInFirefoxParams = params
-            }
-            return true
-        case "deep-link":
-            guard let url = query["url"], Bundle.main.bundleIdentifier == sourceApplication else {
-                break
-            }
-            Router.shared.routeURL(url)
-            return true
-        case "fxa-signin":
-            if AppConstants.MOZ_FXA_DEEP_LINK_FORM_FILL {
-                // FxA form filling requires a `signin` query param and host = fxa-signin
-                // Ex. firefox://fxa-signin?signin=<token>&someQuery=<data>...
-                guard query["signin"] != nil else {
-                    break
-                }
-                let fxaParams: FxALaunchParams
-                fxaParams = FxALaunchParams(query: query)
-                launchFxAFromURL(fxaParams)
-                return true
-            }
-            break
-        default: ()
-        }
-        return false
-    }
-
-    func launchFxAFromURL(_ params: FxALaunchParams) {
-        self.browserViewController.presentSignInViewController(params)
+        handle(nav: routerpath)
+        return true
     }
-
-    func launchFromURL(_ params: LaunchParams) {
-        let isPrivate = params.isPrivate ?? false
-        if let newURL = params.url {
-            self.browserViewController.switchToTabForURLOrOpen(newURL, isPrivate: isPrivate, isPrivileged: false)
-        } else {
-            self.browserViewController.openBlankNewTab(focusLocationField: true, isPrivate: isPrivate)
+    
+    func handle(nav: NavigationPath) {
+        switch nav {
+        case .fxa(let params): NavigationPath.handleFxA(params: params, with: self.browserViewController)
+        case .deepLink(let link): handleDeepLink(link)
+        case .url(let url, let isPrivate): NavigationPath.handleURL(url: url, isPrivate: isPrivate, with: self.browserViewController)
+        }
```

Comment:
This should probably all go in `NavigationRouter` no?

(We should probably treat`AppDelegate.swift` as the next BVC!)

Reply:
Good idea. I’ll move it in there. I think you meant “we _shouldnt_ treat appdelegate as the next bvc right?”